### PR TITLE
Ds/vtk

### PR DESCRIPTION
**Pull request by dsteinmo**
Vtk seems pretty dope.

### CODE REVIEW

File: src/VtkOutputter.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		vtkSmartPointer<vtkUnstructuredGrid> unstructuredGrid = vtkSmartPointer<vtkUnstructuredGrid>::New();
+
+		const char* fieldNameChar = fieldName.c_str();
+		array->SetName(fieldNameChar);
+		array->SetNumberOfValues(Np*K);
+
+		unstructuredGrid->Allocate(K);
+
+		vtkIdType nodes[K][3];
+
```
Comment (k7miller) [inline src/VtkOutputter.cpp:80]: `K` should be a compile-time constant for this to work. I don't believe this is standard C++11.

File: src/VtkOutputter.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		index_type Np = field.rows();
+		index_type nodeId = 0;
+
+		vtkSmartPointer<vtkCellArray> cellArray = vtkSmartPointer<vtkCellArray>::New();
+		vtkSmartPointer<vtkPoints> points = vtkSmartPointer<vtkPoints>::New();
+
+		vtkSmartPointer<vtkDoubleArray> array = vtkSmartPointer<vtkDoubleArray>::New();
+
+		vtkSmartPointer<vtkUnstructuredGrid> unstructuredGrid = vtkSmartPointer<vtkUnstructuredGrid>::New();
+
```
Comment (k7miller) [inline src/VtkOutputter.cpp:70]: Do you need to use `vtkSmartPointer` here? Why not just declare the variables without wrapping them in smart pointers and avoid the extra level of indirection. The code will also be easier to read.

File: src/VtkOutputter.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+			unstructuredGrid->InsertNextCell(VTK_TRIANGLE, 3, nodes[k]);
+		}
+
+		unstructuredGrid->SetPoints(points);
+		unstructuredGrid->GetPointData()->SetScalars(array);
+		unstructuredGrid->GetPointData()->SetActiveScalars(fieldNameChar);
+
+		// Write file
+		vtkSmartPointer<vtkXMLUnstructuredGridWriter> writer =
+			vtkSmartPointer<vtkXMLUnstructuredGridWriter>::New();
```
Comment (k7miller) [inline src/VtkOutputter.cpp:100]: Same comment as above about `vtkSmartPointer`.

File: src/VtkOutputter.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+		unstructuredGrid->Allocate(K);
+
+		vtkIdType nodes[K][3];
+
+		for (index_type k=0; k < K; ++k) {
+			for (index_type n=0; n < Np; ++n) {
+				points->InsertPoint(nodeId, x(n, k), y(n, k), field(n,k) );
+				
+				array->SetValue(nodeId, field(n,k));
```
Comment (k7miller) [inline src/VtkOutputter.cpp:85]: Ideally, `x`, `y`, and `field` are all column-major.

**Review by k7miller — COMMENTED** [review]
The C++ looks pretty good bud, just a few minor comments.

File: src/VtkOutputter.cpp (hunk 2 above)
Comment (dsteinmo) [reply src/VtkOutputter.cpp:70]: For some reason they don't expose public constructors for this. Not sure why they made this choice, but you can only create smart pointers with their New() method.

File: src/VtkOutputter.cpp (hunk 1 above)
Comment (dsteinmo) [reply src/VtkOutputter.cpp:80]: I knew you would say this, yet this is how kitware gives an example of this API on their official wiki. I will see if there is an alternative calling path for the nodes to work around this.

Comment (k7miller) [reply src/VtkOutputter.cpp:80]: I would go with what the C++ standard says and not an example from kitware. From what I have read, GCC allows variable length arrays as a non-standard language extension. I would be surprised if this code built in MSVC.

File: src/VtkOutputter.cpp (hunk 2 above)
Comment (k7miller) [reply src/VtkOutputter.cpp:70]: Fair enough.

File: src/VtkOutputter.cpp (hunk 1 above)
Comment (k7miller) [reply src/VtkOutputter.cpp:80]: Is the `nodes` column bound equal to `Np`? Just wondering if we can drop the magic constant 3.

Comment (dsteinmo) [reply src/VtkOutputter.cpp:80]: The magic number '3' serves as a reminder that the unstructured grid vtk data types don't support triangles with more nodes than 3.

In order to make this work for polynomials higher than one, we'll need to interpolate to a finer mesh (do-able), but out of scope currently.

File: src/VtkOutputter.cpp (hunk 4 above)
Comment (dsteinmo) [reply src/VtkOutputter.cpp:85]: okay, thanks.

**Pull request merged**